Added ability to track location for artwork, and fixed art submission formatting

Braden50 opened the pull request:
Solves #142 

To test:

Try submitting a new artwork (in the process confirm that you can scroll the page and it looks clean)
Once submitted, reload the map page (you need to manually do this) and see if your new artwork shows up in your location.


I had tried this on my laptop and got precise but inaccurate locations. It might just be the fact that I was testing on wifi and it gave the server the location of something further up the network, so let me know how the location works out for you guys.

Braden50 committed:
Added ability to track location for artwork, and fixed art submission formatting

Braden50 committed:
Formatting fix

gwholland3 commented:
I'm getting the same behavior, precise yet inaccurate by 3 or 4 miles. 

A couple other things I noticed that are worth considering:
* If the user doesn't allow their location to be shared, the artwork is given coordinates [0, 0]. Do we want this behavior?
* When I submit a new artwork, using the tab bar to navigate to the map page doesn't show the new artwork. Instead, I have to refresh the page

Braden50 commented:
> I'm getting the same behavior, precise yet inaccurate by 3 or 4 miles.
> 
> A couple other things I noticed that are worth considering:
> 
> * If the user doesn't allow their location to be shared, the artwork is given coordinates [0, 0]. Do we want this behavior?
> * When I submit a new artwork, using the tab bar to navigate to the map page doesn't show the new artwork. Instead, I have to refresh the page

I'm tempted to merge to try it on my phone, but we can see what other people see. 
EDIT: It is possible that it is a tolerance problem as in we aren't given enough significant digits from the API to represent the location precisely (Or at some point it changes from float64 to float32 I don't know).

For the [0, 0] case, I just needed to give it an initial value of the same type. I figured that if someone does not provide a location, we probably shouldn't let them submit an artwork, so the page will only provide the loading spinner and hopefully my convenient message if we can not access their location.

I remember us talking with Bruno once about that refresh issue and John might have said it was a cache thing. I am not sure how to make it re-render when we want.

Braden50 committed:
Added art submission accuracy and view on map capabilities

Braden50 commented:
Update: Just fixed the accuracy issues and added the ability to view specific artworks on the map with "View on map" button. Let me know if it all works!

I wanted to be able to change the initial state (if the lat and long args are not provided) to the user's location, but it seems as though the code in the anon function is not running and I don't know why:

geoControl.on("geolocate", (e) => {
      setUserLocation([e.coords.longitude, e.coords.latitude]);
    });

Waidhoferj commented on frontend/src/pages/ArtSubmission/ArtSubmission.jsx:
```diff
@@ -82,6 +84,35 @@ export default function ArtSubmission() {
+      options);
+  }, [])
+  
+  if (location === [0, 0]) {
+    return (
+      <div>
+        <Spinner absCenter={true} />
+        <h1>Infinite loading? Make sure your location is enabled.</h1>
+      </div>);
+  }
```
Maybe we use the `<ConnectionErrorMessage>` component for this.

Braden50 replied:
Just made the edit. Good suggestion too. I never saw the error, but I imagine that the h1 would look laughably large.

Waidhoferj commented on frontend/src/pages/ArtMap/ArtMap.jsx:
```diff
@@ -116,8 +116,10 @@ export default function ArtMap() {
    */
   function onMapLoad(map) {
     const geoControl = new GeolocateControl({ trackUserLocation: true });
+    if (lat && long) {
+      setMapLocation([lat, long])
+    }
```
I like the idea of viewing the artwork from the map page! If you want the zoom to do the right thing and avoid adding a new route, you could just pass the artwork id. For example `/map/QXJ0d29ya1R5cGU6RmlyZSBXYWxs` will take you to the artwork on the map with that id.

Braden50 replied:
I had initially tried this, but I think it currently just opens up the sidebar for the artwork without changing where the map is looking.

Waidhoferj reviewed (approved):
Looks pretty good! A few comments about optional improvement.

Braden50 committed:
Fixed error message

gwholland3 commented:
@Braden50 I'm still getting the same accuracy issues as before... could it be because of the wifi I'm connected to?

kyleshare commented:
Looks great! It's working well for me but here's a  couple things I noticed.
<img width="1112" alt="Screen Shot 2021-05-31 at 12 38 33 PM" src="https://user-images.githubusercontent.com/52588649/120235291-8ca6bb80-c20e-11eb-88d1-e61db4f43299.png">

- When I click the gallery Icon to upload a picture for a new artwork, I'm taken to this screen instead of having the option to pick files from my computer.
- If you upload multiple artworks at the same location, the "View on map" button will only display the first artwork. However refreshing the page shows all of them.

Braden50 commented:
> @Braden50 I'm still getting the same accuracy issues as before... could it be because of the wifi I'm connected to?

It could be, I just tested it again on my computer with wifi and it worked pretty well. Maybe make sure you got the new edits

Braden50 commented:
> Looks great! It's working well for me but here's a couple things I noticed.
> <img alt="Screen Shot 2021-05-31 at 12 38 33 PM" width="1112" src="https://user-images.githubusercontent.com/52588649/120235291-8ca6bb80-c20e-11eb-88d1-e61db4f43299.png">
> 
> * When I click the gallery Icon to upload a picture for a new artwork, I'm taken to this screen instead of having the option to pick files from my computer.

Small chance you might be pressing the wrong button? The top left upload button is working for me and doesn't open a new page whereas the gallery one does next to the camera icon

> * If you upload multiple artworks at the same location, the "View on map" button will only display the first artwork. However refreshing the page shows all of them.

I am not sure why it would only display one artwork for the location. The "view on map" ability simply routes you to a coordinate on the map where the artwork was said to be. Your error could be part of how the map needs to manually be refreshed.

kyleshare commented:
> Small chance you might be pressing the wrong button? The top left upload button is working for me and doesn't open a new page whereas the gallery one does next to the camera icon
Yeah my mistake I was confusing the gallery Icon for the upload icon.

> I am not sure why it would only display one artwork for the location. The "view on map" ability simply routes you to a coordinate on the map where the artwork was said to be. Your error could be part of how the map needs to manually be refreshed.
Yes refreshing fixes the issue. I thought it was strange that I didn't need to refresh for the first artwork to appear but I did for subsequent ones. But LGTM!

kyleshare commented:
> Small chance you might be pressing the wrong button? The top left upload button is working for me and doesn't open a new page whereas the gallery one does next to the camera icon

Yeah my mistake I was confusing the gallery Icon for the upload icon.

> I am not sure why it would only display one artwork for the location. The "view on map" ability simply routes you to a coordinate on the map where the artwork was said to be. Your error could be part of how the map needs to manually be refreshed.

Yes refreshing fixes the issue. I thought it was strange that I didn't need to refresh for the first artwork to appear but I did for subsequent ones. But LGTM!

gwholland3 reviewed (approved):
> > @Braden50 I'm still getting the same accuracy issues as before... could it be because of the wifi I'm connected to?
> 
> It could be, I just tested it again on my computer with wifi and it worked pretty well. Maybe make sure you got the new edits

Yeah I definitely had the most up-to-date version. If it's working for the rest of you then I say we merge it. Just make sure to rebase though.
